client/src/main/java/org/opensearch/security/client/resources/ResourceSharingNodeClient.java

```diff
@@ -0,0 +1,259 @@
+
+    /**
+     * Checks whether security, resource sharing, or version compatibility is violated.
+     * If so, notifies the listener and returns {@code true}.
+     *
+     * @param contextMessage Context-specific message to append.
+     * @param listener       Listener to notify with an error.
+     * @return {@code true} if feature is disabled or unsupported; otherwise {@code false}.
+     */
+    private boolean handleIfFeatureDisabled(String contextMessage, ActionListener<?> listener) {
+        String reason = getFeatureDisabledReason();
+        if (!reason.isEmpty()) {
+            handleFeatureDisabled(reason + " " + contextMessage, listener);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether security, resource sharing, or version compatibility is violated.
+     * If so, responds with a default value and returns {@code true}.
+     *
+     * @param message         Context-specific log message.
+     * @param listener        Listener to notify with default response.
+     * @param defaultResponse Response to return when feature is disabled.
+     * @param <T>             Type of the default response.
+     * @return {@code true} if feature is disabled or unsupported; otherwise {@code false}.
+     */
+    private <T> boolean handleIfFeatureDisabled(String message, ActionListener<T> listener, T defaultResponse) {
+        String reason = getFeatureDisabledReason();
+        if (!reason.isEmpty()) {
+            log.debug("{} {}", reason, message);
+            listener.onResponse(defaultResponse);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Determines whether the security plugin, resource sharing, or version compatibility
+     * prevents the feature from being used.
+     *
+     * @return A non-empty message if the feature is disabled or unsupported; otherwise empty.
+     */
+    private String getFeatureDisabledReason() {
+        boolean sharingEnabled = settings.getAsBoolean(
+            ConfigConstants.OPENSEARCH_RESOURCE_SHARING_ENABLED,
+            ConfigConstants.OPENSEARCH_RESOURCE_SHARING_ENABLED_DEFAULT
+        );
+
+        Settings securitySettings = settings.getByPrefix(ConfigConstants.SECURITY_SETTINGS_PREFIX);
+        boolean securityDisabled = securitySettings.isEmpty()
+            || settings.getAsBoolean(ConfigConstants.OPENSEARCH_SECURITY_DISABLED, ConfigConstants.OPENSEARCH_SECURITY_DISABLED_DEFAULT);
+
+        if (securityDisabled) return "Security plugin is disabled.";
+        if (!sharingEnabled) return "ShareableResource Access Control feature is disabled.";
+        if (nodeVersion.before(RESOURCE_SHARING_MIN_SUPPORTED_VERSION)) {
+            return "Resource Sharing feature is not supported in version "
+                + nodeVersion
+                + ". Minimum supported version is "
+                + RESOURCE_SHARING_MIN_SUPPORTED_VERSION
+                + ".";
+        }
+
+        return "";
+    }
```

Comment:
I am wondering if we really need this version check here, for the  following reason:

- Plugins always have a dependency on OpenSearch with a specific version
- Plugins must be built newly for each new OpenSearch version. It is not possible to use a plugin which was built for OpenSearch 2.18.0 with OpenSearch 2.19.0.
- Thus, if the Plugin depends on this client of version X and also on other OpenSearch jars of the same version, we already have the information whether the feature is available: If there's a client jar available with this particular version, then the feature will be there.

Reply:
I agree. But this is for safety in case of mixed cluster (i.e. 3.0 handling 2.19 requests). Let me know if you think that this will not be necessary.